Blockchain module cleanup

debris opened the pull request:
changes:

- split `block_to_extras_update` function into 6 smaller functions. Code is cleaner, more maintainable and readable.
- fully separated preparing extas update from applying it to a batch. 
- split `blockchain.rs` into few separate files. This file was becoming a behemoth.
- fixed cleaning up blockchain caches.

debris commented:
I recommend opening `blockchain.rs` file in new tab. Git diff in this case is hardly readable.

gavofyork commented:
needs a rebase. also it's quite big - i presume no logic has been added in this PR?

debris commented:
no, I will rebase

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
-		let mut to_details = to.0.clone();
-		let mut current_from = from.1.clone();
-		let mut current_to = to.1.clone();
+		let mut from_details = self.block_details(&from).expect(&format!("0. Expected to find details for block {:?}", from));
```
this probably shouldn't panic.

debris replied:
Actually here it should. Missing block details means one of two things:

- our db is malformed (or not complete).
- this function is being used incorrectly.

gavofyork replied:
asserting on an external factor such as what is or is not in some particular portion of the database is not particularly defensive coding.

the only proper use of panics/assertions in live code is for specifying logical certainties (`assert!(1 != 0);`) or absolute assertions over the high-level design that are themselves well tested (e.g. triemutdb asserts the root is in the database).

it is sometimes reasonable to use them to specify postconditions when used in order to specify a function's preconditions, this in this case they should be simple enough that it can be done up front in the code prior to the implementation proper.

if you cannot point to comprehensive tests and the structure of the software's design that means your assertion _can never be_ false, then it should not panic, but rather fail-over at a higher level.

in this case, there is a mere "expectation" that the database happens to have a particular entry in it, derived from a potentially arbitrary other number of entries, themselves each populated by code in other modules, and all taken from an argument of unknown provenance.

in short, unless you are willing to bet the farm on this function _never_ being used "improperly", and can provide that by design it cannot happen, then a panic is unacceptable in this position.

debris replied:
I see. I will fix it in next prs.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
-		let mut current_from = from.1.clone();
-		let mut current_to = to.1.clone();
+		let mut from_details = self.block_details(&from).expect(&format!("0. Expected to find details for block {:?}", from));
+		let mut to_details = self.block_details(&to).expect(&format!("1. Expected to find details for block {:?}", to));
```
this probably shouldn't panic.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
 			from_branch.push(current_from);
 			current_from = from_details.parent.clone();
-			from_details = self.block_details(&from_details.parent).expect(&format!("1. Expected to find details for block {:?}", from_details.parent));
+			from_details = self.block_details(&from_details.parent).expect(&format!("2. Expected to find details for block {:?}", from_details.parent));
```
this probably shouldn't panic.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
 			to_branch.push(current_to);
 			current_to = to_details.parent.clone();
-			to_details = self.block_details(&to_details.parent).expect(&format!("2. Expected to find details for block {:?}", to_details.parent));
+			to_details = self.block_details(&to_details.parent).expect(&format!("3. Expected to find details for block {:?}", to_details.parent));
```
this probably shouldn't panic.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -494,11 +383,11 @@ impl BlockChain {
 			from_branch.push(current_from);
 			current_from = from_details.parent.clone();
-			from_details = self.block_details(&from_details.parent).expect(&format!("3. Expected to find details for block {:?}", from_details.parent));
+			from_details = self.block_details(&from_details.parent).expect(&format!("4. Expected to find details for block {:?}", from_details.parent));
```
this probably shouldn't panic.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -494,11 +383,11 @@ impl BlockChain {
 			to_branch.push(current_to);
 			current_to = to_details.parent.clone();
-			to_details = self.block_details(&to_details.parent).expect(&format!("4. Expected to find details for block {:?}", from_details.parent));
+			to_details = self.block_details(&to_details.parent).expect(&format!("5. Expected to find details for block {:?}", from_details.parent));
```
this probably shouldn't panic.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -527,165 +416,225 @@ impl BlockChain {
+			hash: hash,
+			number: number,
+			total_difficulty: total_difficulty,
+			location: match is_new_best {
```
consider rewriting as:
```
if is_new_best {
	// on new best block we need to make sure that all ancestors
	// are moved to "canon chain"
	// find the route between old best block and the new one
	let best_hash = self.best_block_hash();
	let route = self.tree_route(best_hash, parent_hash);

	assert_eq!(number, parent_details.number + 1);

	match route.blocks.len() {
		0 => BlockLocation::CanonChain,
		_ => BlockLocation::BranchBecomingCanonChain {
			ancestor: route.ancestor,
			route: route.blocks.into_iter().skip(route.index).collect()
		}
	}
} else { BlockLocation::Branch }
```
(saves one line and avoids the `true` and `false` symbols).

debris replied:
actually if statement requires more lines, cause I have to wrap it into additional `{ }`

eg.

```rust
BlockInfo {
  hash: hash,
  number: number,
  total_difficulty: total_difficulty,
  location: {
    if is_new_best {
      // ...
    } else {
      // ...
    }
  }
}
```

gavofyork replied:
i think not (your example isn't a fair comparison since you're placing the `if` on its own line whereas `match went on the `location:` line), and requiring the `else` branch to sit on its own line (i would say trivial `else` branches may reasonably be inlined). so now you've got:

```
match is_new_best {
  false => BlockLocation::Branch,
  true => {
    /* 15 lines */
  }
}
```

(20 lines)

versus

```
if is_new_best {
  /* 15 lines */
} else { BlockLocation::Branch }
```

(17 lines)

even if you place the `else` branch on its own line, it's still a line shorter:

```
if is_new_best {
  /* 15 lines */
} else {
  BlockLocation::Branch
}
```

(19 lines)

this way you have 1 fewer indentation levels (elided body is at 1 indent with `if` rather than 2 with `match`) and 4 fewer symbols in the code (`if` + `is_new_best` + `{` + `}` + `else` + `{` + `BlockLocation::Branch` + `}` = 8, versus `match` + `is_new_best` + `{` + `false` + `=>` + `BlockLocation::Branch` + `,` + `true` + `=>` + `{` + `}` + `}` = 12).

`if` wins.

debris replied:
> i think not (your example isn't a fair comparison since you're placing the if on its own line 

I thought that I have to write if inside bracers (like its required to do when if is on the right side of `=>`), but it wasn't necessary here... my bad. Fixed in latest commit.

gavofyork commented on ethcore/src/blockchain/blockchain.rs:
```diff
@@ -527,165 +416,225 @@ impl BlockChain {
-				batch.put_extras(&header.number(), &hash);
+	/// This functions returns modified blocks blooms.
+	///
+	/// TODO: this function needs comprehensive description.
```
maybe do this :)

debris replied:
sure

debris replied:
docs updated :)

gavofyork commented:
it's not especially clear that no logic has been added/changed. maybe add a commentary so we know which functions have been moved where?

given that the tests pass, i'm inclined to let it go this time, but i think for complex refactoring, when it's not clear that it is just repotting, either break down the PRs into two or more smaller refactorings or give clear commentary about what's going on in the PR (e.g. line comments seem to work quite well to guide the reviewer).

arkpar commented:
There should be a test that checks if the blockchain cache has been updated correctly after a fork

gavofyork commented:
can "fixed cleaning up blockchain caches" be placed in a separate PR for ease of review? the other 3 changes this PR makes change no logic.

debris commented:
@gavofyork 
> it's not especially clear that no logic has been added/changed. maybe add a commentary so we know which functions have been moved where?

- moved `blockchain.rs` to `blockchain/blockchain.rs`
- added `BlockInfo` and `BlockLocation` structs (`blockchain/block_info.rs`). They replaced unreadable if statements and comments
- split `block_to_extras_update` into 6 functions. Each of this functions updates only one extras type.
  - `block_info`
  - `prepare_block_hashes_update`
  - `prepare_block_details_update`
  - `prepare_block_receipts_update`
  - `prepare_transaction_addresses_update`
  - `prepare_block_blooms_update`
- moved `BloomIndexer` and it's tests to `blockchain/bloom_indexer.rs`
- moved `BestBlock` to `blockchain/best_block.rs`
- moved `CacheSize` to `blockchain/cache.rs`
- moved `TreeRoute` to blockchain/tree_route.rs`
- moved `ExtrasUpdate` to blockchain/update.rs
- removed `tree_route_aux` function which was not needed (previously used in `block_to_extras_update` function)

@gavofyork 
> can "fixed cleaning up blockchain caches" be placed in a separate PR for ease of review? the other 3 changes this PR makes change no logic.

These are just two lines. [block hashes](https://github.com/ethcore/parity/blob/blockchain_cleanup/ethcore/src/blockchain/blockchain.rs#L445)
and [transaction addresses](https://github.com/ethcore/parity/blob/blockchain_cleanup/ethcore/src/blockchain/blockchain.rs#L463).

@arkpar
> There should be a test that checks if the blockchain cache has been updated correctly after a fork

There are already few tests which check updating block_details && blocks_blooms after a fork. I think it's best to add more test for forking in next prs, cause these tests will make this pr significantly bigger.

gavofyork commented:
ok to merge on my end, though i'd like to see the fork tests and removal of asserts in the not-too-distant future. perhaps introduce an issue to track?

debris commented:
I've added #529 and #530

debris commented:
Started working on sweeping panics on a separate branch. Does `tree_route` function look better? https://github.com/ethcore/parity/commit/8c61131b98a87e29bde585f78aafba98164c77f1#diff-422bab2793f9fb234587f1dadcf690d3R350
